### PR TITLE
DNSName constructor use memchr instead of strchr and cleanup with string_view

### PR DESCRIPTION
**Pull request by darnuria**
Also check length before de-referencing.

## Direct rational for this change:

- Why using `strchr` if we have a length?
- Accepting `char * p` that doesn't contains `\0`

## Is it safe?

- In case of calling from `DNSName` `string` constructor yes `string::length` exclude `\0`
- In case of calling directly it's callee responsibility to provide a valid `length` but it's not less safe than just doing `strchr` IMHO

## Is it still fast?

Not slower in our dev environment.

## Deep rationale: Accepting `char * p` without terminating `\0` 

At work we use https://github.com/PowerDNS/pdns/pull/5160 from @baloo, I work on upstreaming + rebasing this contribution to master (will open an issue to discuss design about it soon).

In a lot of languages `string`s don't terminate with a `\0` like in Rust, so I got some over-read of one with `valgrind` on `strchr` when I wanted to get rid of `string` temporaries just for zeroes.

https://github.com/PowerDNS/pdns/pull/5160/files#diff-8c7f7456f404944349f62aab6fe3444ab12ce90e88d32e3a1a4a93892dd0785aR139-R140

EDIT: small change for clarity English is not my native language

### CODE REVIEW

**Comment by darnuria** [comment]
If it's in scope I can also apply this hunk `length` was added in beb9fad506e4782fa3cd5db0c484b66493943d8b

```diff
diff --git a/pdns/dnsname.cc b/pdns/dnsname.cc
index 4f42350bb..62a843403 100644
--- a/pdns/dnsname.cc
+++ b/pdns/dnsname.cc
@@ -49,9 +49,8 @@ DNSName::DNSName(const char* p, size_t length)
     if(!memchr(p, '\\', length)) {
       unsigned char lenpos=0;
       unsigned char labellen=0;
-      size_t plen=length;
-      const char* const pbegin=p, *pend=p+plen;
-      d_storage.reserve(plen+1);
+      const char* const pbegin=p, *pend = p + length;
+      d_storage.reserve(length + 1);
       for(auto iter = pbegin; iter != pend; ) {
         lenpos = d_storage.size();
         if(*iter=='.')
```

**Review by rgacogne — DISMISSED** [review]
Thanks! The change makes sense to me, and it looks like the tests pass.
I wouldn't mind a second review since this touches a critical function, though :)

**Comment by darnuria** [comment]
No problem even three, totally understand looked at the previous commit it was porcelain edits.

File: pdns/dnsname.cc
Comment (omoerbeek) [reply pdns/dnsname.cc:46]: As a test, I made this constructor private. It does seem to be only called form the other constructors.
- `DNSName(char *p)` uses `strlen()`, so does not handle embedded `\0`s. `length` is s string length.
- `DNSName(const std::string& str) `length` is a string length here as well. Could potentially pass a string with embedded `\0`s. The embedded `\0`s would end up in the labels for the regular case I think. For the backslash case the current test would be incorrect. It is unclear to me what `segmentDNSNameRaw` does with embedded `\0`s,
- The other ct does not call `DNSName(char *, size_t)`.

Comment (rgacogne) [reply pdns/dnsname.cc:46]: I agree the current API is unclear, especially `DNSName(char *p)`. The other constructors are more explicit, you pass the size of the buffer and special characters are supposed to be escaped, otherwise the contract is violated.
I wonder if we could get rid of the `DNSName(char *p)` constructor..

Comment (rgacogne) [reply pdns/dnsname.cc:46]: Note that `strlen` was always called before https://github.com/PowerDNS/pdns/commit/beb9fad506e4782fa3cd5db0c484b66493943d8b

Comment (omoerbeek) [reply pdns/dnsname.cc:46]: Changing it to a buffer length (including terminating '\0' instead of a string length (excluding '\0') would change things I believe, as the end condition of the loop assumes a string length.

Comment (rgacogne) [reply pdns/dnsname.cc:46]: Oh, right, it should be a string length indeed.

Comment (omoerbeek) [reply pdns/dnsname.cc:46]: The handle all calls with `(buf, n)` where `buf` has size `n `the `p[1]` access should be fixed, but the loop should also be double checked if it does not access `buf[n]`, as that would also be an OOB access.

Comment (darnuria) [reply pdns/dnsname.cc:46]: My concern was that by sending `"Hey"` without terminating `\0` (like a rust `Vec<u8>` (continuous `u8` array on the heap) exposed as a cpp `char *`) with length like 3, I ended up with out-of-bound read line 49 the `strchr` because `strchr` care for `\0`.

I am not a c++ expert, no problems if my suggestion does not fit the problem tell me.

An eventual possibility would be to make a constructor with a `string_view` and use (or not) the `DNSName(char *, size_t)  made private constructor. It explicit the intends. From my perspective to pass Rust continuous data of u8 it's perfect.

- It avoids making a string copy just for the `\0` if the `memchr` found a consensus
- It spells out the intends to pass a continuous portion of memory ONLY until length to construct a `DNSName`

Comment (rgacogne) [reply pdns/dnsname.cc:46]: > To handle all calls with (buf, n) where buf has size n the p[1] access should be fixed, but the loop should also be double checked if it does not access buf[n], as that would also be an OOB access.

As far as I can tell the loop is fine, as `iter` is checked against `pend` before being used. I'm not a big fan of using pointer arithmetic for that and would have preferred checking an index, but that should  be OK.

> I am not a c++ expert, no problems if my suggestion does not fit the problem tell me.

I think your use-case is valid, to be clear, and we need to fix/clean up our constructors :)
I 100% agree of taking a `std::string_view` instead of a string, by the way, as it is basically free to construct from a string.

Comment (omoerbeek) [reply pdns/dnsname.cc:46]: > My concern was that by sending `"Hey"` without terminating `\0` (like a rust `Vec<u8>` (continuous `u8` array on the heap) exposed as a cpp `char *`) with length like 3, I ended up with out-of-bound read line 49 the `strchr` because `strchr` care for `\0`.

Yes, that is a real issue. But while reviewing this, a few other issues were spotted.

> I am not a c++ expert, no problems if my suggestion does not fit the problem tell me.
> 
> An eventual possibility would be to make a constructor with a `string_view` and use (or not) the `DNSName(char *, size_t) made private constructor. It explicit the intends. From my perspective to pass Rust continuous data of u8 it's perfect.
> 
>     * It avoids making a string copy just for the `\0` if the `memchr` found a consensus
> 
>     * It spells out the intends to pass a continuous portion of memory ONLY until length to construct a `DNSName`

The handle not-NUL terminated buffers we need to avoid the OOB access with length = 1 and p[0] == '.' something like this: (`p[0]==0 ` is redundant I think)?
```
-  if(length == 0 || p[0]==0 || (p[0]=='.' && p[1]==0)) {
+  if(length == 0 || (length == 1 && p[0]=='.')) {
```

Comment (darnuria) [reply pdns/dnsname.cc:46]: > Yes, that is a real issue. But while reviewing this, a few other issues were spotted.

English is not my native tong I agree with the issues spotted. :)
> The handle not-`NULL` terminated buffers we need to avoid the OOB access with `length = 1` and `p[0] == '.'` something like this: (`p[0] == 0` is redundant I think)?
```diff
-  if(length == 0 || p[0]==0 || (p[0]=='.' && p[1]==0)) {
+ if(length == 0 || (length == 1 && p[0]=='.')) {
```

If it's clear and documented that a length is string length I think so.

But i am a little bit rusty with C/C++, I will take some time in the weekend to test-out, and investigate the `string_view` constructor.

I have no "time pressure" nor critical bug on it, it is for avoiding copies and cleaning a little bit the design of the dlso_backend (may have an issue by monday).

Comment (darnuria) [reply pdns/dnsname.cc:46]: update; Still not have internet in my new apartment yet, so had some issue to contribute this weekend will find some time today. ;)

**Comment by darnuria** [comment]
Update: Have some results with `string_view` may need some caution in review.

If it's better to contribute it in an other pull request let me know! :)

File: pdns/dnsname.cc
Comment (darnuria) [inline pdns/dnsname.cc:47]: It's not the smartest change, but maybe cleaning the internals is worth an other PR, it open a room for improving.

I don't feel confident enough in cpp to do it quickly.

**Comment by darnuria** [comment]
Hi @omoerbeek @rgacogne: small update on this  small update tried to simplify with `string_view` the constructors. For the changes/issues about the internals of the function maybe this weekend or in an evening I might have personal time.

If it's on hold due to other priorities please tell me. :)

**Comment by rgacogne** [comment]
Hi! Thanks for making these changes. The current code looks good to me, but some of the issues raised by Otto regarding the change in behaviour for embedded `\0`s still apply. I think the only way to fix it would be to look at all the places constructing a `DNSName` from a `char*` to check if and how they expect embedded `\0`s to be handled.

**Comment by darnuria** [comment]
For sure, will try to sum them up this afternoon or this weekend, an other possibility but may hurt performance is to search both for `\0` and `\\` explicitly.

Suggestion (not tested yet) : Don't know if it hurt performance (if there a function that takes a lambda it's also possible)
 ```
// Search explicitly for `\\` or an embedded `\0`
if (sw.find_first_of ("\\\0") != std::string::npos) {
/// ....
```

**Comment by darnuria** [comment]
Got caught by time and there is a lot of call to theses constructors maybe I can add some test for this case. also keep the constructor over `char *` that call `strlen` by default what do you think?

File: pdns/dnsname.hh
Comment (darnuria) [inline pdns/dnsname.hh:98]: Maybe keeping this constructor explicitly like it is may avoid some "strange" behaviour in case of "early" `\0`. for `\0` hiden before lenght in string I don't know if it's something expectable.

Comment (darnuria) [reply pdns/dnsname.hh:98]: hmm this may not help *because it fallbacked to the "general `DNSName(char*, size_t) ` but this constructor dealed with "early" `\0` multiple release shipped assuming first `\0` position gives `length` at least with `DNSName(char *)` construction.

Comment (darnuria) [reply pdns/dnsname.hh:98]: Maybe a test like: shall do the trick?

```cpp
BOOST_AUTO_TEST_CASE(test_early_null) {
  const char * test = "powerdns\0.com.";
  DNSName data(test);
  DNSName expected("powerdns");
  BOOST_CHECK_EQUAL(data, expected);
}
```

Comment (darnuria) [reply pdns/dnsname.hh:98]: this test added in `pdns/test-dnsname_cc.cc` passed both on this branch and on master ff43a0c2cbe44a7d0233d0c0446801efca1fc3ce just added the case in the middle of the file look like it runned had the number of test going from 237 to 238 with it.

Comment (darnuria) [reply pdns/dnsname.hh:98]: (had no employer pressure it can wait just had some idea to test it, figuring out if the early `\0` is already checked etc)

**Comment by omoerbeek** [comment]
Did a rebase and used a ref for the `string_view` arg plus a few other minor changes.

I reviewed all existing callers: there are no callers of `DNSName(const char *, size_t) with a length passed in any of our existing code, except in `dnsname.hh` where it fills in `strlen()`. That means that all calls pass through the existing `strlen()` for size case. That means that embedded NULs will be seen as end-of-string.

So it is safe to do this modification to `string_view`, as it will not change the semantics existing callers expect, the `string_view` ct will do the same as before. 

*if* you pass an explicit length in the future embdded NULs might end up in your `DNSName`.

**Comment by omoerbeek** [comment]
I'm approving this. An extra review is needed, as this is fundamental.

**Review by omoerbeek — APPROVED** [review]
See latest comments: another review by somebody else needed.

File: pdns/dnsname.hh
Comment (rgacogne) [inline pdns/dnsname.hh:102]: I'm not sure it makes sense in that case to pass views by reference instead of passing them by value. Views are cheap to create and copy, and if I'm not mistaken requiring a reference here means we can only pass lvalues.

**Review by rgacogne — APPROVED** [review]
Looks good to me! One nit about the use of a reference instead of a value, but I'm fine with merging this PR anyway.

File: pdns/dnsname.hh
Comment (omoerbeek) [reply pdns/dnsname.hh:102]: Anything larger than a pointer is "large" imo :-). I'll revert that part.

Comment (darnuria) [reply pdns/dnsname.hh]: Thanks it was instructive, I went for a ref due to my Rust background like ` &str` (slice type on string) but since string_view are "like slice type" i should have passed it directly.

**Pull request merged**
